source/objects/lander.lua

```diff
@@ -0,0 +1,579 @@
+        fun.AddScreen("Settings")
+    end
+
+	-- Rest angle
+	if math.max(lander.angle) > 360 then lander.angle = 0 end
+	
+	-- Update ship
+    moveShip(lander, dt)
+    updateSmoke(dt)
+    playSoundEffects(lander)
+    checkForContact(lander, dt)
+end
+
+
+
+function Lander.draw(worldOffset)
+	-- draw the lander and flame
+	for landerId, lander in ipairs(garrLanders) do
+		local sx, sy = 1.5, 1.5
+		local drawingX = lander.x - worldOffset
+		local drawingY = lander.y
+
+		-- fade other landers in multiplayer mode
+		if landerId == 1 then
+			love.graphics.setColor(1,1,1,1)
+		else
+			love.graphics.setColor(1,1,1,0.5)
+		end
+
+		local ox = lander.width / 2
+		local oy = lander.height / 2
+		love.graphics.draw(garrImages[5], drawingX,drawingY, math.rad(lander.angle), sx, sy, ox, oy)
+
+		--[[
+			FIXME:
+			It would be better to avoid creating these variables every tick. This will likely
+			be resolved with further code improvements in the future.
+		--]]
+		-- draw flames
+		local flameSprite	= garrImages[4]
+		local flameWidth	= flameSprite:getWidth()
+		local flameHeight	= flameSprite:getHeight()
+		local ox 			= flameWidth / 2
+		local oy 			= flameHeight / 2
+
+		if lander.engineOn then
+			local angle = math.rad(lander.angle)
+			love.graphics.draw(flameSprite, drawingX, drawingY, angle, sx, sy, ox, oy)
+			lander.engineOn = false
+		end
+		if lander.leftEngineOn then
+			local angle = math.rad(lander.angle + 90)
+			love.graphics.draw(flameSprite, drawingX, drawingY, angle, sx, sy, ox, oy)
+			lander.leftEngineOn = false
+		end
+		if lander.rightEngineOn then
+			local angle = math.rad(lander.angle - 90)
+			love.graphics.draw(flameSprite, drawingX, drawingY, angle, sx, sy, ox, oy)
+			lander.rightEngineOn = false
+		end
+
+		-- draw smoke trail
+		for _, smoke in pairs(garrSmokeSprites) do
+			-- FIXME: All images / frames should have a width/height variable to avoid hardcoded numbers!
+			-- FIXME: Smoke related stuff should be in it's own local function
+			local drawingX = smoke.x - worldOffset
+			local drawingY = smoke.y
+			local spriteId = cf.round(smoke.dt)
+			if spriteId < 1 then spriteId = 1 end
+			-- not sure why the smoke sprite needs to be rotate +135. Suspect the image is drawn wrong. This works but!
+			love.graphics.draw(gSmokeSheet,gSmokeImages[spriteId], drawingX, drawingY, math.rad(lander.angle + 135))
+		end
+
+		-- draw label
+		love.graphics.setNewFont(10)
+		local offsetX, offsetY = 14, 10
+		love.graphics.print(lander.name, drawingX + offsetX, drawingY - offsetY)
+		love.graphics.setColor(1,1,1,1)
+	end
+end
+
+
+
+function Lander.keypressed(key, scancode, isrepeat)
+	local lander = garrLanders[1]
+	-- 2 = base type (fuel)
+	if Lander.isOnLandingPad(lander, 2) then
+		if key == "1" then
+			buyThrusters(lander)
+		elseif key == "2" then
+			buyLargeTank(lander)
+		elseif key == "3" then
+			buyRangefinder(lander)
+		elseif key == "4" then
+			buySideThrusters(lander)
+		end
+	end
+end
+
+
+return Lander
```

Comment:
Needs a new line here at the end

Reply:
Changed 48b3b7989fbf344ab2bb512b1a4dc71ed0666b9c

Reply:
Hm, the new line automatically gets removed for me. I mean, I don't really care about it. vscode shows it non-the-less.